connectors/sync_job_runner.py

```diff
@@ -516,19 +522,46 @@ async def generator(self):
                 raise UnsupportedJobType
 
     async def update_ingestion_stats(self, interval):
+        last_cursor = None
         while True:
             await asyncio.sleep(interval)
 
             if not await self.reload_sync_job():
                 break
 
+            cursor = (
+                None
+                if not self.data_provider  # If we failed before initializing the data provider, we don't need to change the checkpoint
+                else (
+                    self.data_provider.sync_cursor()
+                    if self.sync_job.is_content_sync()
+                    else None
+                )
+            )
+
+            if cursor != last_cursor:
+                self.sync_job.log_debug(
+                    "Connector reported a new cursor, triggering batch flush before saving"
+                )
+                last_cursor = cursor
+                # Fun stuff happens here - this will block until the data is flushed
+                # to Elasticsearch. Once it's flushed, we can continue and save the job state
+                # This is done to avoid data inconsistencies - if we don't flush, then
+                # we can store cursors before the data that preceded this cursor was ingested.
+                await self.sync_orchestrator.trigger_flush()
+                self.sync_job.log_debug(
+                    "Data was successfully flushed before saving the cursor"
+                )
+
             result = self.sync_orchestrator.ingestion_stats()
             ingestion_stats = {
                 INDEXED_DOCUMENT_COUNT: result.get(INDEXED_DOCUMENT_COUNT, 0),
                 INDEXED_DOCUMENT_VOLUME: result.get(INDEXED_DOCUMENT_VOLUME, 0),
                 DELETED_DOCUMENT_COUNT: result.get(DELETED_DOCUMENT_COUNT, 0),
             }
-            await self.sync_job.update_metadata(ingestion_stats=ingestion_stats)
+            await self.sync_job.update_metadata(
+                ingestion_stats=ingestion_stats, cursor=cursor
```

Comment:
Do we also need to check here that the cursor was updated but non-null?

Reply:
`update_metadata` internally checks [that](https://github.com/elastic/connectors/pull/2879/files#diff-c2f73fe85d1652beb4af8939b73360a399a3496c1863e34b900e12d694420571R402-R403): 

```
                if cursor:
                    doc["connector"] = {"sync_cursor": cursor}
```

Idk if it would cause confusion too, maybe there should be a way for the connector to indicate that cursor needs a reset, e.g. if it became corrupted?